### PR TITLE
QUIC Connection State Machine and Front End I/O API Implementation

### PR DESCRIPTION
**Pull request by hlandau**
This PR has been structured by commit to ease review; it is recommended to review it by commit.

## Refactors to other components

Other components have been refactored slightly where necessary, such as the record layer. The RX depacketiser has been fleshed out to complete implementation.

## Dummy handshake protocol & successful application E2E test

This PR was tested by implementing a dummy handshake protocol with psuedo-ClientHello, psuedo-ServerHello, etc. Most of these messages are empty except for ClientHello and EncryptedExtensions, which contain QUIC transport parameters only. The same traffic secret is always negotiated. An existing QUIC implementation was modified to use this dummy handshake layer. A simple echo server implementation was made using this implementation. A connection was successfully made to this server using this PR, application data was sent on a single bidirectional stream, and the same data as echoed back from the echo server was successfully read by the application, using SSL_read and SSL_write calls.

[For reference, the test program which was used for this can be found here.](https://gist.github.com/hlandau/cce896c86943beff80ac853cb0c17c28)

## Handshake layer API

Thus, this does not yet implement standard QUIC, but as soon as the real handshake layer is available we can just drop it in. (@mattcaswell: you may wish to look at the API to the dummy handshake layer used here (`quic_dummy_handshake.h`). Of course, when we use the real handshake layer, I assume it will be an SSL object and be interfaced somewhat differently, but this should give some idea of what is needed.)

## Testing

This PR does not include tests. There are several reasons for this:

- Since it uses a dummy handshake layer, it cannot be tested against standard QUIC implementations.
- Since we do not have any kind of server support yet, we have nothing to test connectivity with unless we depend on third party services on the internet, or on third party implementations pulled in during testing, even notwithstanding the above.
- The CSM cannot really be tested in isolation and to test the CSM is to test the entire implementation, which really needs to be a dedicated effort.

## API Changes

Previously, the expected API for QUIC usage was determined via the demo-driven design process. During implementation, some desirable changes have been discovered. In particular:

- The DDD demos did not reflect the fact that SSL objects can traditionally be used with distinct read and write BIO. The proposed new API `SSL_get_poll_fd` has been split into `SSL_get_poll_rfd` and `SSL_get_poll_wfd`;
- The DDD demos took return codes from `SSL_get_error` as a cue as to whether they should poll for POLLIN, POLLOUT, or both. This is so horribly fiddly and error prone that I now view it as a very bad idea. Added two calls: `SSL_want_net_read` and `SSL_want_net_write`. As usual, `SSL_get_tick_timeout` returns the recommended polling timeout value.
- Added `SSL_tick`.
- Non-blocking I/O is always used internally. If an FD is given to the QUIC implementation it will set it as non-blocking. This is discussed further in a comment in `quic_impl.c`.
- Moreover, to my knowledge, SSL objects generally haven't been told whether they are being used with a blocking socket or not and just deal with EWOULDBLOCK etc. if it happens. This doesn't really work for QUIC where we always use non-blocking I/O internally (see comment in `quic_impl.c` for discussions on the reasons for this). Moreover we cannot find out if a socket is blocking when it is given to us because Windows has no API for getting the blocking flag, only setting it (...). Thus a new API is introduced to configure the blocking mode for QUIC connections: `SSL_set_blocking_mode(SSL *s, int blocking)` and `SSL_get_blocking_mode()`.

Looking forward to people ripping this apart &mdash; if there's anything I missed, let me know.

### CODE REVIEW

**Comment by kroeckx** [comment]
Since I think this will now let me test some simple things, I've modified the client fuzzer like this:
```c
--- a/fuzz/client.c
+++ b/fuzz/client.c
@@ -11,6 +11,7 @@
 #include <time.h>
 #include <openssl/rand.h>
 #include <openssl/ssl.h>
+#include <openssl/quic.h>
 #include <openssl/rsa.h>
 #include <openssl/dsa.h>
 #include <openssl/ec.h>
@@ -64,7 +65,7 @@ int FuzzerTestOneInput(const uint8_t *buf, size_t len)
         return 0;
 
     /* This only fuzzes the initial flow from the client so far. */
-    ctx = SSL_CTX_new(SSLv23_method());
+    ctx = SSL_CTX_new(OSSL_QUIC_client_method());
     if (ctx == NULL)
         goto end;
 
```

But it fails on the following 2 lines:
    OPENSSL_assert(SSL_set_min_proto_version(client, 0) == 1);
    OPENSSL_assert(SSL_set_cipher_list(client, "ALL:eNULL:@SECLEVEL=0") == 1);

Commenting those 2 lines let's it continue, but doesn't actually produce any coverage at all, all input files seem to produce the same coverage.

**Comment by paulidale** [comment]
I am a bit skeptical about forcing sockets into non-blocking mode.  One of the requirements is that applications have the capability to control the event loop themselves and changing the socket mode underneath is contrary to this.  I'm also not entire sure we should be working at the FD level.  It's quite normal for an application to handle networking itself and rely on BIOs for the TLS part.  Supporting the same approach for QUIC seems desirable, if not mandatory.

For MVP, we might get away with this but I'm not sure it's a longer term solution.

**Comment by hlandau** [comment]
>I am a bit skeptical about forcing sockets into non-blocking mode. One of the requirements is that applications have the capability to control the event loop themselves and changing the socket mode underneath is contrary to this. I'm also not entire sure we should be working at the FD level.

I'm open to suggestions as to how we can do things otherwise.

>It's quite normal for an application to handle networking itself and rely on BIOs for the TLS part. Supporting the same approach for QUIC seems desirable, if not mandatory.

It's always been explicitly intended that we support an application using a memory buffer BIO, hence the work on BIO_dgram_pair. This is not yet supported in this PR but it will need to happen.

**Comment by mattcaswell** [comment]
> It's always been explicitly intended that we support an application using a memory buffer BIO, hence the work on BIO_dgram_pair. This is not yet supported in this PR but it will need to happen.

Or any other custom BIO...that the application wishes to use.

File: Configure [hunk 1]
```diff
@@ -167,7 +167,8 @@ my @gcc_devteam_warn = qw(
     -Wsign-compare
     -Wshadow
     -Wformat
-    -Wtype-limits
+    -Wno-type-limits
+    -Wno-tautological-constant-out-of-range-compare
```
[the remark this answers is not in the document]
Comment (hlandau) [reply Configure:171]: Namely things like this:
```c
if (SIZE_MAX > UINT64_MAX)
```
This has been causing a few problems recently. We keep committing code that breaks on 32-bit platforms where `SIZE_MAX != UINT64_MAX` and then fixing it later (see @levitte's PR to fix some of the QUIC code on Windows). Testing against `SIZE_MAX` and `UINT64_MAX` is the right/portable way to do this, but some compilers like to complain about a comparison that can "never" be true (e.g. on 64-bit platforms).

I think we should disable these warnings on the grounds that they're effectively actively encouraging one to write worse code.

File: ssl/quic/quic_record_rx.c [hunk 2]
```diff
@@ -40,8 +40,9 @@ static ossl_inline int pkt_is_marked(const uint64_t *bitf, size_t pkt_idx)
 typedef struct rxe_st RXE;
 
 struct rxe_st {
+    OSSL_QRX_PKT        pkt;
     RXE                *prev, *next;
-    size_t              data_len, alloc_len;
+    size_t              data_len, alloc_len, refcount;
```
[the remark this answers is not in the document]
Comment (hlandau) [reply ssl/quic/quic_record_rx.c:45]: Yeah. I avoided this API on the grounds that it requires allocating a lock which would be quite annoying in this circumstance. Nothing here is thread safe yet, so it's not needed here.

**Comment by hlandau** [comment]
@mattcaswell The first commit in 'this' PR is [QUIC ACKM: Add support for psuedo-loss](https://github.com/openssl/openssl/pull/19582/commits/e09660f23f599556ea5ee48988c6e9c1fd546548).

File: Configure (hunk 1 above)
[the remark this answers is not in the document]
Comment (mattcaswell) [reply Configure:171]: Wouldn't that specific code be better written as a preprocessor conditional?

```` c
#if UINT64_MAX > SIZE_MAX
````

File: ssl/quic/quic_record_rx.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattcaswell) [reply ssl/quic/quic_record_rx.c:45]: Fair enough

File: ssl/quic/quic_impl.c [hunk 3]
```diff
@@ -9,179 +9,2104 @@
...
+        ossl_quic_stream_map_update_state(&qc->qsm, qc->stream0);
+
+        /*
+         * Try and send.
+         *
+         * TODO(QUIC): It is probably inefficient to try and do this
+         * immediately, plus we should eventually consider Nagle's algorithm
+         * etc.
+         */
+        reactor_tick(&qc->rtor);
```
Comment (mattcaswell) [inline ssl/quic/quic_impl.c:2483]: The logic here looks unexpected to me. If we're in blocking mode we shouldn't return until everything is written. But here we don't even check if we're in blocking mode if we've written at least one byte, and then we just return. If we're not in blocking mode I expect SSL_write to fail if we have failed to write *all* the data (and SSL_get_error() should return SSL_ERROR_WANT_WRITE) - unless SSL_MODE_ENABLE_PARTIAL_WRITE has been set.

In the case of non-blocking mode and SSL_MODE_ENABLE_PARTIAL_WRITE has not been set, then we if we *did* do a partial write then we will need to remember how much we wrote, and continue from where we left off next time we are called.

File: ssl/quic/quic_impl.c [hunk 4]
```diff
@@ -9,179 +9,2104 @@
...
+         * freed up.
+         */
+        args.qc         = qc;
+        args.buf        = buf;
+        args.len        = len;
+        args.written    = written;
+
+        res = reactor_block_until_pred(&qc->rtor, quic_write_again, &args, 0);
+        if (res <= 0)
+            return 0;
```
Comment (mattcaswell) [inline ssl/quic/quic_impl.c:2497]: Shouldn't we loop and try and append the data again?

File: ssl/quic/quic_impl.c [hunk 5]
```diff
@@ -9,179 +9,2104 @@
...
+ * wfd_want_write=0 is not the same as passing -1 for both FDs.
+ *
+ * deadline is a timestamp to return at. If it is ossl_time_infinite(), the call
+ * never times out.
+ *
+ * Returns 0 on error and 1 on success. Timeout expiry is considered a success
+ * condition. We don't elaborate our return values here because the way we are
+ * actually using this doesn't currently care.
+ */
+static int poll_two_fds(int rfd, int rfd_want_read,
```
Comment (mattcaswell) [inline ssl/quic/quic_impl.c:240]: I am an uneasy about this function and surprised to see us doing select/poll in the library. Is that what we decided we needed to do when we designed the API? I forget now.

File: ssl/quic/quic_record_shared.c [hunk 6]
```diff
@@ -199,8 +199,9 @@ int ossl_qrl_enc_level_set_provide_secret(OSSL_QRL_ENC_LEVEL_SET *els,
     if (el == NULL
         || md_name == NULL
         || init_key_phase_bit > 1 || is_tx < 0 || is_tx > 1
-        || (init_key_phase_bit > 0 && enc_level != QUIC_ENC_LEVEL_1RTT))
+        || (init_key_phase_bit > 0 && enc_level != QUIC_ENC_LEVEL_1RTT)) {
         return 0;
+    }
```
Comment (mattcaswell) [inline ssl/quic/quic_record_shared.c:204]: Why this change?

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Still more review to do. But here's a bunch more comments.

File: Configure (hunk 1 above)
[the remark this answers is not in the document]
Comment (hlandau) [reply Configure:171]: Unfortunately this is not possible because we define these constants in `numbers.h` using an expression based on `sizeof()`. We're actually polyfilling these ourselves and not using the values provided by the OS.

File: ssl/quic/quic_impl.c [hunk 7]
```diff
@@ -1587,14 +1593,71 @@ int ossl_quic_clear(SSL *s)
...
+ *
+ * Handling the different BIOs is difficult:
+ *
+ *   - It is more or less a requirement that we use non-blocking network I/O;
+ *     we need to be able to have timeouts on recv() calls, and make best effort
+ *     (non blocking) send() and recv() calls.
+ *
+ *     The only sensible way to do this is to configure the socket into
+ *     non-blocking mode. We could try to do select() before calling send() or
+ *     recv() to get a guarantee tha the call will not block, but this will
```
Comment (mattcaswell) [inline ssl/quic/quic_impl.c:1985]: typo: that the call

File: ssl/quic/quic_impl.c [hunk 8]
```diff
@@ -1230,16 +1298,29 @@ static void csm_tick(QUIC_TICK_RESULT *res, void *arg)
...
+        res->want_net_write = 0;
+        res->tick_deadline  = ossl_time_infinite();
+        return;
+    }
+
     deadline    = ossl_ackm_get_loss_detection_deadline(qc->ackm);
     if (ossl_time_compare(now, deadline) >= 0)
         ossl_ackm_on_timeout(qc->ackm);
 
+
```
Comment (mattcaswell) [inline ssl/quic/quic_impl.c]: Too many blank lines here

File: ssl/quic/quic_impl.c [hunk 9]
```diff
@@ -1555,32 +1668,182 @@ int ossl_quic_conn_on_handshake_confirmed(QUIC_CONNECTION *qc)
...
+ */
+void ossl_quic_conn_raise_protocol_error(QUIC_CONNECTION *qc,
+                                         uint64_t error_code,
+                                         uint64_t frame_type,
+                                         const char *reason)
+{
+    QUIC_TERMINATE_CAUSE tcause = {0};
+
+    tcause.error_code = error_code;
+    tcause.frame_type = frame_type;
```
Comment (mattcaswell) [inline ssl/quic/quic_impl.c:1811]: Shouldn't we call `ERR_raise` somewhere in this function? If so we should consider how we pass through the originating line number data.

File: test/recipes/80-test_ssl_new.t [hunk 10]
```diff
@@ -128,7 +128,6 @@ my %skip = (
   "25-cipher.cnf" => disabled("ec") || disabled("tls1_2"),
   "26-tls13_client_auth.cnf" => disabled("tls1_3") || ($no_ec && $no_dh),
   "29-dtls-sctp-label-bug.cnf" => disabled("sctp") || disabled("sock"),
-  "31-quic.cnf" => $no_quic || $no_ec,
```
Comment (mattcaswell) [inline test/recipes/80-test_ssl_new.t:131]: Why are we deleting this?

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Final comments from my first pass through this PR below.

The CI failures are relevant. Most importantly API docs are missing.

File: ssl/quic/quic_impl.c [hunk 11]
```diff
@@ -9,179 +9,2104 @@
...
+            //   break;
+            // QUIC_TPARAM_STATELESS_RESET_TOKEN
+            // QUIC_TPARAM_MAX_UDP_PAYLOAD_SIZE
+            // QUIC_TPARAM_INITIAL_MAX_STREAMS_BIDI
+            // QUIC_TPARAM_INITIAL_MAX_STREAMS_UNI
+            // QUIC_TPARAM_ACK_DELAY_EXPONENT
+            // QUIC_TPARAM_MAX_ACK_DELAY
+            // QUIC_TPARAM_DISABLE_ACTIVE_MIGRATION
+            // QUIC_TPARAM_PREFERRED_ADDR
+            // QUIC_TPARAM_ACTIVE_CONN_ID_LIMIT
```
[the remark this answers is not in the document]
Comment (hlandau) [reply ssl/quic/quic_impl.c]: Superceded by subsequent commit

File: ssl/quic/quic_impl.c (hunk 4 above)
Comment (hlandau) [reply ssl/quic/quic_impl.c:2497]: This is what `reactor_block_until_pred` and `quic_write_again` does?

File: ssl/quic/quic_impl.c [hunk 12]
```diff
@@ -9,179 +9,2104 @@
...
-        OPENSSL_free(qc);
-        ssl = NULL;
+    /* Initialise the QUIC_CONNECTION's stub header. */
+    ssl_base = &qc->ssl;
+    if (!ossl_ssl_init(ssl_base, ctx, SSL_TYPE_QUIC_CONNECTION)) {
+        ssl_base = NULL;
         goto err;
     }
-    qc->tls = ossl_ssl_connection_new(ctx);
-    if (qc->tls == NULL || (sc = SSL_CONNECTION_FROM_SSL(qc->tls)) == NULL)
```
[the remark this answers is not in the document]
Comment (hlandau) [reply ssl/quic/quic_impl.c:31]: It will be readded. It was not needed for now.

File: ssl/quic/quic_impl.c [hunk 13]
```diff
@@ -9,179 +9,2104 @@
...
+{
+    /*
+     * Send a packet, if we need to. Best effort. The TXP consults the CC and
+     * applies any limitations imposed by it, so we don't need to do it here.
+     *
+     * Best effort. In particular if TXP fails for some reason we should still
+     * flush any queued packets which we already generated.
+     */
+    if (ossl_quic_tx_packetiser_generate(qc->txp,
+                                         TX_PACKETISER_ARCHETYPE_NORMAL) == 2)
```
[the remark this answers is not in the document]
Comment (hlandau) [reply ssl/quic/quic_impl.c:1491]: Refactored this in the TXP. I'll fix this when I rebase on top of it.

File: ssl/quic/quic_impl.c [hunk 14]
```diff
@@ -9,179 +9,2104 @@
...
+}
+
+/*
+ * ossl_quic_conn_on_protocol_error {{{2
+ * --------------------------------
+ */
+void ossl_quic_conn_on_protocol_error(QUIC_CONNECTION *qc, uint64_t error_code,
+                                      const char *reason)
+{
+    fprintf(stderr, "# PROTOCOL ERROR: %lu: %s\n", error_code,
```
[the remark this answers is not in the document]
Comment (hlandau) [reply ssl/quic/quic_impl.c]: Removed in a subsequent commit

File: ssl/quic/quic_impl.c [hunk 15]
```diff
@@ -9,179 +9,2104 @@
...
+    csm_discard_el(qc, QUIC_ENC_LEVEL_HANDSHAKE);
+    qc->handshake_confirmed = 1;
+    return 1;
+}
+
+/*
+ * ossl_quic_conn_on_protocol_error {{{2
+ * --------------------------------
+ */
+void ossl_quic_conn_on_protocol_error(QUIC_CONNECTION *qc, uint64_t error_code,
```
[the remark this answers is not in the document]
Comment (hlandau) [reply ssl/quic/quic_impl.c]: Renamed to `ossl_quic_conn_raise_protocol_error` and used in subsequent commits

File: ssl/quic/quic_impl.c [hunk 16]
```diff
@@ -9,179 +9,2104 @@
...
+ * QUIC Front-End I/O API: Network BIO Configuration {{{1
+ * =================================================
+ */
+
+/*
+ * csm_determine_poll_fd_from_bio {{{2
+ * ------------------------------
+ * Determine what FD we should poll on based on our underlying network BIO.
+ */
+static int csm_determine_poll_fd_from_bio(QUIC_CONNECTION *qc, BIO *net_bio, int *pfd)
```
[the remark this answers is not in the document]
Comment (hlandau) [reply ssl/quic/quic_impl.c]: Going to be refactored based on the updated design document.

File: test/quicapitest.c [hunk 17]
```diff
@@ -22,56 +22,6 @@ static OSSL_PROVIDER *defctxnull = NULL;
...
-            goto end;
-    }
-
-    ret = 1;
-
- end:
-    SSL_free(serverquic);
-    SSL_free(clientquic);
-    SSL_CTX_free(sctx);
-    SSL_CTX_free(cctx);
```
[the remark this answers is not in the document]
Comment (hlandau) [reply test/quicapitest.c:70]: We don't have a server to talk to currently and don't have a real handshake layer, so I don't see how (for now).

[the remark this answers is not in the document]
Comment (mattcaswell) [reply test/quicapitest.c:70]: Perhaps `#if 0` it out and add a `TODO(QUIC)` to fix it, rather than deleting it? When we have a test server implementation it could be reintroduced.

File: test/recipes/80-test_ssl_new.t (hunk 10 above)
Comment (mattcaswell) [reply test/recipes/80-test_ssl_new.t:131]: Comment out and leave a TODO?

**Comment by hlandau** [comment]
Rebased. Updated to implement BIO poll descriptors and the new API in the design document.

More updates to come.

File: ssl/quic/quic_impl.c (hunk 3 above)
Comment (hlandau) [reply ssl/quic/quic_impl.c:2497]: This behaviour of SSL_write is highly surprising.

As an aside, are our our own API docs accurate? SSL_write(3) talks about this, but BIO_write(3) says nothing about partial writes not being allowed. But BIO_write(3) for BIO_f_ssl is a simple wrapper around SSL_write(3). BIO_write(3) doesn't say anything about having to pass the same buffer again. So the BIO docs seem to be quite misleading here.

Comment (mattcaswell) [reply ssl/quic/quic_impl.c:2497]: I don't think BIO_write() generally has this restriction but SSL_write does. BIO_write() for BIO_f_ssl() will just never do a partial write I think - which is still within the API contract for BIO_write().

File: Configure (hunk 1 above)
[the remark this answers is not in the document]
Comment (mattcaswell) [reply Configure:171]: On compile error I start seeing this note at the end of the compilation errors with this change:

````
cc1: note: unrecognized command-line option '-Wno-tautological-constant-out-of-range-compare' may have been intended to silence earlier diagnostics
````

If there are no compile errors it just seems to ignore the unrecognised flag and continue anyway.

**Comment by hlandau** [comment]
Updated.

The first new commit is "QUIC CSM: Application-level error indications".

Changes:
- Better error indications
- Support SSL_MODE_ENABLE_PARTIAL_WRITE={0,1} and model the semantics correctly
- Flesh out some TLS-related API calls
- Fix calculation of deadlines
- Implement SSL_pending
- Implement blocking SSL_connect
- Refactor connection process to work in terms of SSL_do_handshake
- Add documentation for new APIs
- Support third party BIO poll descriptors

**Comment by mattcaswell** [comment]
Added "tests:deferred" label because we really need better testing of this - but can't really do it until we have some mechanism for a test server within the test harness.

**Comment by hlandau** [comment]
Rebased. Minor changes, starting with "QUIC CSM: Correctly change DCID on first Initial packet".

**Comment by hlandau** [comment]
After working on implementation of the test server I've come to the conclusion that the best way forward will be to substantially refactor this PR. Otherwise, implementation of the test server would end up duplicating large amounts of the code in this PR. Part of the problem is that the CSM is fused with the front-end I/O API, but we don't have a public API for server functionality yet, thus there is a need to separate the two. (I ended up writing it like this because at the time I didn't have a good idea of how the two should be separated, but the way to do this is now much clearer.)

In particular, I intend to refactor this PR to separate what I will call the "libssl API Personality Layer" (i.e., the front-end I/O API), and client-specific functionality, from the majority of the CSM code. Basically it will happen like this:
- A new kind of object, a `QUIC_CHANNEL` will be defined. This is basically the same as a QUIC connection, it just has to be called something else because we already have something called a `QUIC_CONNECTION`. This will contain most of the state currently contained in the `QUIC_CONNECTION` but will not be an SSL object and will not implement the libssl API personality layer, or anything specific to client or server roles.
- `QUIC_CONNECTION` will be refactored to operate in terms of a `QUIC_CHANNEL`. Client-specific functionality as well as the libssl API personality layer will be implemented here. For example, `QUIC_CONNECTION` will still be responsible for e.g. implementing SSL object modes (partial write, etc.), blocking semantics, etc. (`QUIC_CHANNEL` will be non-blocking only.)
- There will be an internal test server which also is implemented in terms of a `QUIC_CHANNEL` but which is not an SSL object.

Once this refactor is done I expect test server functionality to "fall out" of it almost for free, so this will be the bulk of the work for enabling test server functionality.

**Comment by hlandau** [comment]
This PR has been superceded by #19703.
